core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -2007,12 +2008,47 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       "offsets.commit.required.acks must be greater or equal -1 and less or equal to offsets.topic.replication.factor")
     require(BrokerCompressionCodec.isValid(compressionType), "compression.type : " + compressionType + " is not valid." +
       " Valid options are " + BrokerCompressionCodec.brokerCompressionOptions.mkString(","))
-    require(!processRoles.contains(ControllerRole) || controllerListeners.nonEmpty,
-      s"${KafkaConfig.ControllerListenerNamesProp} cannot be empty if the server has the controller role")
+    if (usesSelfManagedQuorum) {
+      require(controlPlaneListenerName.isEmpty,
+        s"${KafkaConfig.ControlPlaneListenerNameProp} is not supported in KRaft mode.")
+      if (processRoles.contains(ControllerRole)) {
+        // has controller role (and optionally broker role as well)
+        // controller.listener.names must be non-empty
+        // every one must appear in listeners
+        // each port appearing in controller.quorum.voters must match the port in exactly one controller listener
+        require(controllerListeners.nonEmpty,
+          s"${KafkaConfig.ControllerListenerNamesProp} must contain at least one value appearing in the '${KafkaConfig.ListenersProp}' configuration when running the KRaft controller role")
+        val listenerNameValues = listeners.map(_.listenerName.value).toSet
+        require(controllerListenerNames.forall(cln => listenerNameValues.contains(cln)),
+          s"${KafkaConfig.ControllerListenerNamesProp} must only contain values appearing in the '${KafkaConfig.ListenersProp}' configuration when running the KRaft controller role")
+        RaftConfig.parseVoterConnections(quorumVoters).asScala.foreach { case (nodeId, addressSpec) =>
```

Comment:
This checks the port of every voter. We should only check the port of this voter/controller. Instead of `foreach` maybe `get(nodeId)`? This must be `Some` because I think we have another validation check for the id of this controller is in the voter set.

Reply:
> We should only check the port of this voter/controller

I coded it that way first, but upon further thought I came to the conclusion that we should check every voter -- not just this one.  Based on my comment above, I believe this should actually require the port to match the first controller listener's port, and that further adds weight to checking every node.

Reply:
Hmm... what happens for the case of running multiple controllers on a single box?  Maybe we do only check this node so that the other ones can have a different port just for this situation?  It does seem to me to be a corner case, that normally all the ports should match.

Reply:
> Hmm... what happens for the case of running multiple controllers on a single box? Maybe we do only check this node so that the other ones can have a different port just for this situation? It does seem to me to be a corner case, that normally all the ports should match.
 
This case is very common in tests, integration tests and system tests.